lib/livebook/runtime/node_pool.ex

```diff
@@ -0,0 +1,84 @@
+defmodule Livebook.Runtime.NodePool do
+  use GenServer, async: false
+
+  # A pool for reusing child node names.
+  #
+  # `pool` refers to the list of unused names
+  #
+  # `buffer_time` refers to time the pool waits before
+  # adding a name to `pool`,  which by default is 1 minute.
+
+  @default_time 6_000
+
+  # Client interface
+
+  @doc """
+  Starts the node pool with the buffer time as 1 minute.
+  """
+  def start do
+    GenServer.start(__MODULE__, @default_time, name: __MODULE__)
+  end
+
+  @doc """
+  Starts the node pool with the buffer time as input.
+  """
+  def start(buffer_time) do
+    GenServer.start(__MODULE__, buffer_time, name: __MODULE__)
+  end
+
+  @doc """
+  Retuns a node name.
+
+  Generates a new one if node is empty, or takes on from pool.
+  """
+  def get_name(parent) do
+    GenServer.call(__MODULE__, {:get_name, parent})
+  end
+
+  # Server side code
+
+  @impl GenServer
+  def init(buffer_time) do
+    :net_kernel.monitor_nodes(true)
```

Comment:
You want to pass `node_type: :all` as option, as the nodes we start are hidden.

Reply:
Why wouldn't we set it as hidden nodes only ?

Reply:
We can... but it doesn't hurt to track all in case we remove the --hidden option in the future. :) And given that we will track which names we own and which names we gave.